### PR TITLE
tp-qemu: Add driver_install script for windows guest.

### PR DESCRIPTION
**Pull request by CongSmile**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1293206

### CODE REVIEW

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    if not os.path.exists("C:\Windows\system32\certutil.exe"):
+        shutil.copy("%s\certutil.exe" % vol_utils, "C:\Windows\system32")
+    if not os.path.exists("C:\Windows\system32\certadm.dll"):
+        shutil.copy("%s\certadm.dll" % vol_utils, "C:\Windows\system32")
+    try:
+        os.system(certutil_cmd)
+    except Exception, err:
+        print "Can not install the certificate with cmd %s!" % certutil_cmd
+        print "Details refers: %s" % err
+        exit_status = -1
```
Comment (spcui) [inline qemu/deps/win_driver_install/win_driver_install.py]: 1 win_driver_install.py|43 col 1 error| W0612 local variable 'exit_status' is assigned to but never used [pyflakes]                                                 
  2 win_driver_install.py|56 col 1 error| W0612 local variable 'exit_status' is assigned to but never used [pyflakes]
  3 win_driver_install.py|73 col 1 error| E0602 undefined name 'install_driver_cmd' [pyflakes]
  4 win_driver_install.py|75 col 1 error| W0612 local variable 'exit_status' is assigned to but never used [pyflakes]
  5 win_driver_install.py|109 col 1 error| W0612 local variable 'exit_status' is assigned to but never used [pyflakes]

Comment (CongSmile) [inline qemu/deps/win_driver_install/win_driver_install.py]: It's used in __main__

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
+import os
+import re
+import sys
+import shutil
+
+
+# Usage of win_driver_install.py.
+def usage(script_name):
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: use import argparse to write your usage and options

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    use_str = " %s" % script_name
+    use_str += " --op={install_driver|uninstall_driver|query_driver|help}"
+    use_str += " --driver_path= --driver_name="
+    use_str += " --device_name= --vol_utils="
+    print use_str
+
+
+# Get dpinst.exe
+def getdpinst(vol_utils):
+    get_proc_arch_cmd = "echo %PROCESSOR_ARCHITECTURE%"
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: you can get this kind of infor from sys module

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                        os.system(uninstall_driver_cmd)
+                    except Exception, err:
+                        print "Driver uninstallat failed with cmd %s!" % uninstall_driver_cmd
+                        print "Details refers: %s and dpinst.log" % err
+                        exit_status = -1
+
+
+# Get the specific current driver version
+def get_current_driver_ver(device_name):
+    get_driver_ver_cmd = "wmic path win32_pnpsigneddriver where Devicename='%s' get driverversion" % device_name
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: so long

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    key = "\d*\.\d*\.\d*\.\d*"
+    driver_version = os.popen(get_driver_ver_cmd).read()
+    return re.findall(key, driver_version, re.M)[-1]
+
+
+# Get expected driver version
+def get_expected_driver_ver(driver_path):
+    key = "\d*\.\d*\.\d*\.\d*"
+    for file in os.listdir(driver_path):
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: avoid to use python keywords

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        exit_status = -1
+
+
+if __name__ == "__main__":
+    if len(sys.argv) < 6:
+        script_name = sys.argv[0]
+        usage(script_name)
+        sys.exit(-1)
+
+    for argv in sys.argv[1:]:
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: see comments1

File: qemu/deps/win_driver_install/win_driver_install.py (hunk 4 above)
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: os.walk or os.path.walk will helpful to do it.

File: qemu/deps/win_driver_install/win_driver_install.py (hunk 6 above)
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: error_msg = ("xxxx"
                      "yyyy")
looks more beautiful.

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        verify_driver_ver(driver_path, device_name)
+    except Exception, err:
+        print "Driver installation failed with cmd %s!" % install_driver_cmd
+        print "Details refers: %s and dpinst.log" % err
+        exit_status = -1
+
+
+# Uninstall the specific driver
+def uninstall_driver(driver_name):
+    driver_store = "C:\Windows\System32\DriverStore\FileRepository"
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: is it a fix path for x86 and x86_64 windows guest

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    uninstall_driver_cmd = "C:\dpinst.exe /U %s /C /LM /Q /F"
```
Comment (suqinhuang) [inline qemu/deps/win_driver_install/win_driver_install.py]: Could you check if there is driver folder first? the path may be updated

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+import re
+import sys
+import shutil
+
+
+# Usage of win_driver_install.py.
+def usage(script_name):
+    print "Usage of %s:" % script_name
+    use_str = " %s" % script_name
+    use_str += " --op={install_driver|uninstall_driver|query_driver|help}"
```
Comment (suqinhuang) [inline qemu/deps/win_driver_install/win_driver_install.py]: suggest to short the option, -i, -u ...

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    key = "\d*\.\d*\.\d*\.\d*"
+    for file in os.listdir(driver_path):
+        if ".inf" in file:
+            file_path = os.path.join(driver_path, file)
+            inf_file = open(file_path)
+            driver_info = inf_file.readlines()
+            inf_file.close()
+            for line in driver_info:
+                if "DriverVer" in line:
+                    return re.findall(key, line, re.M)[-1]
```
Comment (suqinhuang) [inline qemu/deps/win_driver_install/win_driver_install.py]: it's better to conver to int, there maybe blank after the string e.g "1000" != "1000    "

File: qemu/deps/win_driver_install/win_driver_install.py (hunk 9 above)
Comment (CongSmile) [inline qemu/deps/win_driver_install/win_driver_install.py]: @suqinhuang 

Is this better ?

# python win_driver_install.py --help
usage: win_driver_install.py [-h] [-i] [-u] [-q] [--driver_path DRIVER_PATH]
                             [--driver_name DRIVER_NAME]
                             [--device_name DEVICE_NAME]
                             [--vol_utils VOL_UTILS]

Windows Driver Operation

optional arguments:
  -h, --help            show this help message and exit
  -i, --install_driver  operation for install driver
  -u, --uninstall_driver
                        operation for uninstall driver
  -q, --query_driver    operation for query driver
  --driver_path DRIVER_PATH
                        driver path
  --driver_name DRIVER_NAME
                        driver name
  --device_name DEVICE_NAME
                        the corresponding device name with driver
  --vol_utils VOL_UTILS
                        volume of WIN_UTILS

File: qemu/deps/win_driver_install/win_driver_install.py (hunk 7 above)
Comment (CongSmile) [inline qemu/deps/win_driver_install/win_driver_install.py]: Yes.

**Comment by CongSmile** [comment]
Hi @suqinhuang and @xutian 

Could you help review it again?

I have updated the code according to your comments.

Thanks.

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+# Install certificate
+def certutil(vol_utils):
+    certutil_cmd = "certutil -addstore -f TrustedPublisher %s\redhat.cer" % vol_utils
+    if not os.path.exists("C:\Windows\system32\certutil.exe"):
+        shutil.copy("%s\certutil.exe" % vol_utils, "C:\Windows\system32")
+    if not os.path.exists("C:\Windows\system32\certadm.dll"):
+        shutil.copy("%s\certadm.dll" % vol_utils, "C:\Windows\system32")
+    try:
+        os.system(certutil_cmd)
```
Comment (suqinhuang) [inline qemu/deps/win_driver_install/win_driver_install.py]: suggest to print the step log. eg "Install certifacation" and cmd log

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+        dpinstdir = "%s\dpinst_32.exe" % vol_utils
+    if not os.path.exists("C:\dpinst.exe"):
+        shutil.copy(dpinstdir, "C:\dpinst.exe")
+    else:
+        print "dpinst.exe is already existed"
+
+
+# Install certificate
+def certutil(vol_utils):
+    certutil_cmd = "certutil -addstore -f TrustedPublisher %s\redhat.cer" % vol_utils
```
Comment (suqinhuang) [inline qemu/deps/win_driver_install/win_driver_install.py]: certutil or certutil.exe

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    except Exception, err:
+        error_msg = "Can not install certificate with cmd %s!\n" % certutil_cmd
+        error_msg += "Details refers: %s" % err
+        print error_msg
+        exit_status = -1
+
+
+# Install the specific driver
+def install_driver(driver_path, device_name, vol_utils):
+    install_driver_cmd = "C:\dpinst.exe /A /PATH %s /C /LM /Q /F" % driver_path
```
Comment (suqinhuang) [inline qemu/deps/win_driver_install/win_driver_install.py]: The same as above, add step description and print log

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    key = "\d*\.\d*\.\d*\.\d*"
+    get_driver_ver_cmd = "wmic path win32_pnpsigneddriver where" \
+                         " Devicename='%s' get driverversion" % device_name
+    driver_version = os.popen(get_driver_ver_cmd).read()
+    return re.findall(key, driver_version, re.M)[-1].strip()
+
+
+# Get expected driver version
+def get_expected_driver_ver(driver_path):
+    key = "\d*\.\d*\.\d*\.\d*"
```
Comment (suqinhuang) [inline qemu/deps/win_driver_install/win_driver_install.py]: Why not get .inf file like line 54
inf_file = "%s.inf" % driver_name

File: qemu/deps/win_driver_install/win_driver_install.py (hunk 12 above)
Comment (CongSmile) [inline qemu/deps/win_driver_install/win_driver_install.py]: They are same, certutil = certutil.exe

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    for line in driver_info:
+        if "DriverVer" in line:
+            return re.findall(key, line, re.M)[-1].strip()
+
+
+# Verify the installed driver is same as expected.
+def verify_driver_ver(driver_path, device_name):
+    current_driver_ver = get_current_driver_ver(device_name)
+    expected_driver_ver = get_expected_driver_ver(driver_path)
+    print "Compare driver version is same as expected."
```
Comment (suqinhuang) [inline qemu/deps/win_driver_install/win_driver_install.py]: make more clear ? "Compare current driver version with expected one"

**Comment by suqinhuang** [comment]
Acked-by: Suqin Huang

**Comment by CongSmile** [comment]
For some guests, file extensions is hided as default, e.g. ***.inf 

Need update this part.

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+# Install certificate
+def certutil(vol_utils):
+    certutil_cmd = "certutil -addstore -f TrustedPublisher %s\\redhat.cer" % vol_utils
+    if not os.path.exists("C:\Windows\system32\certutil.exe"):
+        shutil.copy("%s\certutil.exe" % vol_utils, "C:\Windows\system32")
+    if not os.path.exists("C:\Windows\system32\certadm.dll"):
+        shutil.copy("%s\certadm.dll" % vol_utils, "C:\Windows\system32")
+    try:
+        print "Install certificate !"
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: os.system will not raise any exception here

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+        exit_status = -1
+
+
+# Uninstall the specific driver
+def uninstall_driver(driver_name):
+    driver_store = "C:\Windows\System32\DriverStore\FileRepository"
+    uninstall_driver_cmd = "C:\dpinst.exe /U %s /C /LM /Q /D"
+    inf_file = "%s.inf" % driver_name
+    try:
+        os.path.exists(driver_store)
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: here too

**Comment by CongSmile** [comment]
Hi @xutian 
I have updated the code according to your comments, could you please help review it again?

Thanks.

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+        print "dpinst.exe is already existed"
+    if not os.path.exists("C:\dpinst.xml"):
+        dpinst_xml = "%s\dpinst.xml" % vol_utils
+        shutil.copy(dpinst_xml, "C:\dpinst.xml")
+    else:
+        print "dpinst.xml is already existed"
+
+
+# Install certificate
+def certutil(vol_utils):
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: All path in your code should be a raw string, so I think you need "r" to ask a raw string path.

Thanks,
xu

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    certutil_cmd = "certutil -addstore -f TrustedPublisher %s\\redhat.cer" % vol_utils
+    if not os.path.exists("C:\Windows\system32\certutil.exe"):
+        shutil.copy("%s\certutil.exe" % vol_utils, "C:\Windows\system32")
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: I'm not sure that we always have permission to copy file in  system root directory.

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 20]
```diff
@@ -0,0 +1,171 @@
+import os
+import re
+import sys
+import shutil
+import argparse
+import subprocess
+
+
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: SYSTEM_ROOT = "C:\Windows\System32"

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+    parser.add_argument('--driver_path',
+                        help='driver path', dest='driver_path', action='store')
+    parser.add_argument('--driver_name',
+                        help='driver name', dest='driver_name', action='store')
+    parser.add_argument('--device_name',
+                        help='the corresponding device name with driver',
+                        dest='device_name', action='store')
+    parser.add_argument('--vol_utils',
+                        help='volume of WIN_UTILS',
+                        dest='vol_utils', action='store')
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: add option "-o | --output"  to recording stderr and stdout to log

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+    get_driver_ver_cmd = "wmic path win32_pnpsigneddriver where" \
+                         " Devicename='%s' get driverversion" % device_name
+    driver_version = os.popen(get_driver_ver_cmd).read()
+    return re.findall(key, driver_version, re.M)[-1].strip()
+
+
+# Get expected driver version
+def get_expected_driver_ver(driver_path, driver_name):
+    key = "\d*\.\d*\.\d*\.\d*"
+    inf_file = get_inf_files(driver_path, driver_name)
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: type  of inf_file is list, so I guess that  below lines will cause OSError Exception.

Comment (CongSmile) [inline qemu/deps/win_driver_install/win_driver_install.py]: Yes, have converted the list to string in line 100.

**Comment by CongSmile** [comment]
@xutian 

The code has been updated, could you help review it again?
Thanks.

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+        error_msg += "Current driver version %s" % current_driver_ver
+        error_msg += " is not equal to the expected %s." % expected_driver_ver
+        logger.error(error_msg)
+        sys.exit(1)
+    logger.info("Current driver version %s is same as expected." % current_driver_ver)
+
+
+# Print log.
+def show_log_output(result_file):
+    fp = open(result_file, 'r')
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: with open(result_file, 'r') as fd:
       print  os.linesep.join(fd.readlines())

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+                        help='driver name', dest='driver_name', action='store')
+    parser.add_argument('--device_name',
+                        help='the corresponding device name with driver',
+                        dest='device_name', action='store')
+    parser.add_argument('--vol_utils',
+                        help='volume of WIN_UTILS',
+                        dest='vol_utils', action='store')
+    arguments = parser.parse_args()
+
+    result_file = "c:\driver_install.log"
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: all path  should be raw string!!

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+            return re.findall(key, line, re.M)[-1].strip()
+
+
+# Verify the installed driver is same as expected.
+def verify_driver_ver(driver_path, device_name, driver_name):
+    current_driver_ver = get_current_driver_ver(device_name)
+    expected_driver_ver = get_expected_driver_ver(driver_path, driver_name)
+    logger.info("Compare whether driver version is same as expected.")
+    if current_driver_ver != expected_driver_ver:
+        error_msg = "Driver installation failed !\n"
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: better to use  '()' instead '+='

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+                         " Devicename='%s' get driverversion" % device_name
+    driver_version = os.popen(get_driver_ver_cmd).read()
+    return re.findall(key, driver_version, re.M)[-1].strip()
+
+
+# Get expected driver version
+def get_expected_driver_ver(driver_path, driver_name):
+    key = "\d*\.\d*\.\d*\.\d*"
+    inf_file = get_inf_files(driver_path, driver_name)
+    inf_file = open("".join(inf_file))
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: same as line#121

File: qemu/deps/win_driver_install/win_driver_install.py [hunk 27]
```diff
@@ -0,0 +1,190 @@
...
+    driver_info = inf_file.readlines()
+    inf_file.close()
+    for line in driver_info:
```
Comment (xutian) [inline qemu/deps/win_driver_install/win_driver_install.py]: info = inf_file.read()
re.findall(r'.*DriverVer.*(key)', info, re.M)[-1].strip()

**Comment by xutian** [comment]
still leak docstring

**Comment by CongSmile** [comment]
@xutian Could you help review it again?
Thanks.

**Comment by CongSmile** [comment]
@xutian, code has been updated, could you help review it again?
Thanks.

**Comment by CongSmile** [comment]
@xutian, code has been updated, could you help review it again?
Thanks.

**Comment by xutian** [comment]
looks ok to me.

**Comment by suqinhuang** [comment]
Acked-by: Suqin Huang

**Pull request merged**